### PR TITLE
WIP: Bump Prometheus-operator and node_exporter to latest versions

### PR DESCRIPTION
**Pull request by lilic**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
While writing kube-state-metrics perf tests, I noticed that the version of Prometheus operator and node_exporter are out of date, so I decided to contribute. Let me know if I overstepped the line, happy to close the PR! 

I tried to keep the overriding changes as close to as they were, let me know if I missed something important?

I also added node_exporter DaemonSet manifest instead of the pod one, but happy to revert the changes if you had a reason for one pod!

And switched replicas of Prometheus to 2, but since I do not know the full purpose of this Prometheus instance, maybe having 1 replica is enough in your case? (e.g. the trafeoffs of consuming less memory in a scale and perf test vs availability of Prometheus monitoring might be indeed a normal thing)

**Special notes for your reviewer**:
(Note I am the maintainer of both Prometheus operator and kube-prometheus which the original manifests are based on)

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by mm4tt** [comment]
The PR is marked as WIP, let me know when it's ready for review.

**Comment by lilic** [comment]
@mm4tt Thanks for reply! Just wanted to first get initial feedback if this is something you all want? If yes, can you maybe addressed the comments in the description. After that I can address them and unmark it as WIP. Hope that makes sense? :) Thank you!

File: clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml [hunk 1]
```diff
@@ -1,31 +1,34 @@
...
+        app.kubernetes.io/version: 0.45.0
     spec:
       containers:
       - args:
-        - --logtostderr=true
-        - --config-reloader-image=gcr.io/k8s-testimages/quay.io/coreos/configmap-reload:v0.0.1
-        - --prometheus-config-reloader=gcr.io/k8s-testimages/quay.io/coreos/prometheus-config-reloader:v0.30.0
-        {{if $PROMETHEUS_SCRAPE_KUBELETS}}
         - --kubelet-service=kube-system/kubelet
-        {{end}}
```
Comment (paulfantom) [inline clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml:27]: Seems like some templating was removed.

**Comment by mm4tt** [comment]
I took a cursory glance. A few notes:

- node-exporter has been updated to the newest version in https://github.com/kubernetes/perf-tests/pull/1665
- node-exporter needs to be kept as a pod - it's installed as a static pod on master nodes
- prometheus doesn't need two replicas in our setup, we've never faced any loss of prometheus availability in our scale tests. On the other hand, having two instances comes with a cost, more resources needs, more api-calls, we'd have to change some of our custom logic (e.g. we have a custom logic for dumping prometheus databases after a test), etc. TL;DR, the gains of having 2+ replicas don't justify the costs here

It'd be great if you can update the prometheus operator and dependencies to the newest version. It's something that we'll have to do eventually anyway. But, could you keep the change as minimalistic as possible? I see that you changed/added a lot of things, e.g. a lot of new grafana dashboards, that aren't really needed. Also, please make sure to honor our custom overrides and settings (e.g. the thing Pawel caught). 

Thanks!

File: clusterloader2/pkg/prometheus/manifests/prometheus-clusterRole.yaml [hunk 2]
```diff
@@ -1,19 +1,19 @@
...
   - ""
   resources:
   - nodes/metrics
-  - services
-  - endpoints
-  - pods
   verbs:
   - get
-  - list
-  - watch
```
Comment (marseel) [inline clusterloader2/pkg/prometheus/manifests/prometheus-clusterRole.yaml:16]: Are you sure we can remove this? 
If I am not mistaken when we are using service/pod monitors prometheus is doing list and watch for these two objects, right?

File: clusterloader2/pkg/prometheus/manifests/prometheus-prometheus.yaml [hunk 3]
```diff
@@ -56,10 +72,10 @@ spec:
...
-  {{if $PROMETHEUS_SCRAPE_WINDOWS_NODES}}
-  additionalScrapeConfigs:
-    name: windows-scrape-configs
-    key: windows-scrape-configs.yaml
+  query:	
+    maxSamples: 100000000	
+  {{if $PROMETHEUS_SCRAPE_WINDOWS_NODES}}	
+  additionalScrapeConfigs:	
+    name: windows-scrape-configs	
+    key: windows-scrape-configs.yaml
```
Comment (marseel) [inline clusterloader2/pkg/prometheus/manifests/prometheus-prometheus.yaml:80]: nit: remove whitespaces

File: clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml [hunk 4]
```diff
@@ -1,31 +1,34 @@
...
-        image: gcr.io/k8s-testimages/quay.io/coreos/prometheus-operator:v0.31.0
+        - --prometheus-config-reloader=gcr.io/k8s-testimages/quay.io/prometheus-operator/prometheus-config-reloader:v0.45.0
+        image: gcr.io/k8s-testimages/quay.io/prometheus-operator/prometheus-operator:v0.45.0
```
Comment (marseel) [inline clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml:31]: FTR these images do not exist right now, we will need to push them manually.

File: clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml [hunk 5]
```diff
@@ -39,7 +42,27 @@ spec:
             memory: 100Mi
         securityContext:
           allowPrivilegeEscalation: false
-          readOnlyRootFilesystem: true
+      - args:
+        - --logtostderr
+        - --secure-listen-address=:8443
+        - --tls-cipher-suites=TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305
+        - --upstream=http://127.0.0.1:8080/
+        image: quay.io/brancz/kube-rbac-proxy:v0.8.0
```
Comment (marseel) [inline clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml:50]: I am not sure what this container is doing, but either way if we want to use it then we will need to mirror it to gcr (in the past we had issues, where our tests were failing, because quay.io was down.

**Comment by marseel** [comment]
I took quick look. Definitely it's something that we want ;) 
I also agree with @mm4tt that it's better not to increase number of replicas for now. It would be safer and easier to review if we would split it into smaller PRs (as @mm4tt mentioned)

File: clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml (hunk 5 above)
Comment (paulfantom) [reply clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml:50]: Just a note: There are also [talks in kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/issues/1785) to make automatic cloning of kube-rbac-proxy. Plus we (kube-rbac-proxy maintainers) want to donate that to k8s in the future when we find time to actually do it.

**Comment by lilic** [comment]
Thank you all!! Just heads up I did not forget about this :)

**Comment by s-urbaniak** [comment]
side note: to spread up the work load, I am taking this over from @lilic :-)

File: clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml (hunk 1 above)
Comment (s-urbaniak) [reply clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml:27]: added back in, indeed, the variable is declared at the top :+1:

File: clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml (hunk 4 above)
Comment (s-urbaniak) [reply clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml:31]: ah, i see. Can you advise how we can get this image updated?

File: clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml (hunk 5 above)
Comment (s-urbaniak) [reply clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml:50]: it seems 0.8.0 is available as per https://github.com/kubernetes-sigs/kubebuilder/pull/1955, updating this PR accordingly.

Comment (s-urbaniak) [reply clusterloader2/pkg/prometheus/manifests/0prometheus-operator-deployment.yaml:50]: @marseel: I am new to this ecosystem, but can `gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0` be referenced her as well or can you advise how I can help in having a `gcr.io/k8s-testimages/quay.io/brancz/kube-rbac-proxy:v0.8.0`?

tldr: kube-rbac-proxy is protects endpoints by executing a token and subject access review against k8s api to validate if service account token in flight has permissions to query metrics.

**Comment by s-urbaniak** [comment]
@mm4tt @marseel let me actually submit separate PRs for the suggested changes so it is easier to track :+1: side note: i am also prometheus-operator and kube-prometheus maintainer, hence it might look like big changes here but they were more or less taken 1:1 from kube-prometheus upstream which evolved gradually.

re: grafana dashboards: do you use explicit custom grafana dashboards? Otherwise we could indeed vendor them 1:1 to catch up with upstream changes from kube-prometheus.

**Comment by mm4tt** [comment]
> re: grafana dashboards: do you use explicit custom grafana dashboards?

Yes, we have our own grafana dashboards defined in https://github.com/kubernetes/perf-tests/tree/97fedf5c484e6438c2c13854bb901241493377ba/clusterloader2/pkg/prometheus/manifests/dashboards

**Comment by lilic** [comment]
Thanks Serg 🎉
